File: Problems/Dynamic-programming/Longest-Common-Subsequence/Longest-Common-Subsequence.cs
using System;

// Using tabulation - Time: O(m*n)

public class Solution

    /*
    public int LongestCommonSubsequence(string text1, string text2) {
        
        int m = text1.Length;
        int n = text2.Length;

        if(m < n)
        {
            (m, n) = (n, m);
            (text1, text2) = (text2, text1);
        }  

        int[] dp = new int[n + 1];

        for(int i = 1; i <= m; i++)
        {
            int prev = 0;
            for(int j = 1; j <= n; j++)
            {
                int temp = dp[j];
                if(text1[i - 1] == text2[j - 1]) {
                    dp[j] = prev + 1;
                } else {
                    dp[j] = Math.Max(dp[j], dp[j-1]);
                }
                prev = temp;
            }
        }  

        return dp[n];
    }

    public int LongestCommonSubsequence(string s1, string s2) {
        
        int [,] dp = new int[s1.Length + 1, s2.Length + 1];
        for(int i = 1; i <= s1.Length; i++) {
            for(int j = 1; j <= s2.Length; j++) 
            {
                if(s1[i-1] == s2[j-1]) {
                    dp[i, j] = dp[i-1, j - 1]+1;
                }
                else{
                    dp[i, j] = Math.Max(dp[i-1, j], dp[i, j-1]);
                }
            }
        }
        return dp[s1.Length, s2.Length];
    }

    public class Solution {
    public int LongestCommonSubsequence(string str1, string str2) {
        int m = str1.Length;
        int n = str2.Length;

        if(m < n) {
            (m, n) = (n, m);
            (str1, str2) = (str2, str1);
        }

        int[] prev = new int[n+1];
        int[] curr = new int[n+1];

        for(int i = 1; i <= m; i++) {
            for(int j = 1; j <= n; j++ ) {
                   if (str1[i - 1] == str2[j - 1]) {
                // Characters match: Extend LCS
                    curr[j] = prev[j - 1] + 1;
                } else {
                // No match: Take the max from previous row or left cell
                    curr[j] = Math.Max(prev[j], curr[j - 1]);
                }
            }
             // Swap references instead of copying
            var temp = prev;
            prev = curr;
            curr = temp;
        }
            return prev[n];
    }

}
    */
{
    public int LongestCommonSubsequence(string text1, string text2)
    {
        int m = text1.Length;
        int n = text2.Length;

        if (m < n)
        {
            (m, n) = (n, m);
            (text1, text2) = (text2, text1);
        }

        int[] dp = new int[n + 1];

        for (int i = 0; i < m; i++)
        {
            int prev = 0;

            for (int j = 0; j < n; j++)
            {

                int cur = dp[j + 1];
                if (text1[i] == text2[j])
                {
                    dp[j + 1] = prev + 1;
                }
                else
                {
                    dp[j + 1] = Math.Max(dp[j], dp[j + 1]);
                }
                prev = cur;
            }
        }

        return dp[n];
    }
}

class Program
{
    static void Main()
    {
        string text1 = "abcde";
        string text2 = "ace";

        Console.WriteLine("Input: text1 = " + text1 + ", text2 = " + text2);

        Solution sol = new Solution();
        int result = sol.LongestCommonSubsequence(text1, text2);

        Console.WriteLine("Output: " + result);
    }
}

File: Problems/Dynamic-programming/Longest-Common-Subsequence.js
/**
 * Finds the length of the longest common subsequence (LCS) between two strings.
 * Uses dynamic programming to build a table of LCS lengths.
 *
 * @param {string} str1 - First input string.
 * @param {string} str2 - Second input string.
 * @return {number} - Length of the LCS.
 * Time complexity: O(m.n). The solution involves filling a grid of size m x n, where m
 *  is the length of str1 and n is the length of str2.
Space complexity: O(m.n). The dpGrid requires space proportional to the product of the
 lengths of the two strings.
 * 
 */
export default function longestCommonSubsequence(str1, str2) {
    // Create a 2D array (dp) with (str1.length + 1) rows and (str2.length + 1) columns
    // Initialize all values to 0
    var dp = Array.from({ length: str1.length + 1 }, () => new Array(str2.length + 1).fill(0));

    // Build the DP table
    for (var i = 1; i <= str1.length; i++) {
        for (var j = 1; j <= str2.length; j++) {
            if (str1[i - 1] === str2[j - 1]) {
                // Characters match: extend LCS from previous state
                dp[i][j] = dp[i - 1][j - 1] + 1;
            } else {
                // No match: take the max LCS by excluding one character
                dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
            }
        }
    }

    // The final answer is in the bottom-right corner of the table
    return dp[str1.length][str2.length];
}

/**
 * @param {string} str1
 * @param {string} str2
 * @return {number}
 * Time complexity: O(m.n). Two nested loops are used: one iterating over the 
 * length of str1 (m) and the other iterating over the length of str2 (n).
Space complexity: O(n). Only two rows, each of size n+1, are used for computation,
 reducing space usage from O(m.n) to O(n).
 */
export default function longestCommonSubsequence(str1, str2) {
    // Ensure str2 is the shorter string to minimize space usage
    // Ensure str2 is the shorter string to minimize space usage
    if (str1.length < str2.length) {
        [str1, str2] = [str2, str1];
    }

    const m = str1.length, n = str2.length;

    // Create two 1D arrays to store current and previous row results
    let prev = new Array(n + 1).fill(0);
    let curr = new Array(n + 1).fill(0);

    for (let i = 1; i <= m; i++) {
        for (let j = 1; j <= n; j++) {
            if (str1[i - 1] === str2[j - 1]) {
                // Characters match: Extend LCS
                curr[j] = prev[j - 1] + 1;
            } else {
                // No match: Take the max from previous row or left cell
                curr[j] = Math.max(prev[j], curr[j - 1]);
            }
        }
        // Update previous row after processing the current row
        prev = [...curr]; // No need to swap, just copy the current row to previous row
    }

    // The LCS length is in the last cell of the current row (which is now the prev row)
    return curr[n];
}